profile/filter_test.go

```diff
@@ -0,0 +1,252 @@
+		},
+		{
+			hide: regexp.MustCompile("lib.so"),
+			fm:   true,
+			hm:   true,
+		},
+		{
+			show: regexp.MustCompile("foo.c"),
+			fm:   true,
+			hnm:  true,
+		},
+		{
+			show: regexp.MustCompile("notfound"),
+			fm:   true,
+		},
+	} {
+		prof := *testProfile1.Copy()
+		gf, gi, gh, gnh := prof.FilterSamplesByName(tc.focus, tc.ignore, tc.hide, tc.show)
+		if gf != tc.fm {
+			t.Errorf("Filter #%d, got fm=%v, want %v", tx, gf, tc.fm)
+		}
+		if gi != tc.im {
+			t.Errorf("Filter #%d, got im=%v, want %v", tx, gi, tc.im)
+		}
+		if gh != tc.hm {
+			t.Errorf("Filter #%d, got hm=%v, want %v", tx, gh, tc.hm)
+		}
+		if gnh != tc.hnm {
+			t.Errorf("Filter #%d, got hnm=%v, want %v", tx, gnh, tc.hnm)
+		}
+	}
+}
+
+func TestFilter2(t *testing.T) {
+	mappings := []*Mapping{
+		{ID: 1, Start: 0x10000, Limit: 0x40000, File: "map0", HasFunctions: true, HasFilenames: true, HasLineNumbers: true, HasInlineFrames: true},
+		{ID: 2, Start: 0x50000, Limit: 0x70000, File: "map1", HasFunctions: true, HasFilenames: true, HasLineNumbers: true, HasInlineFrames: true},
+	}
+	functions := []*Function{
+		{ID: 1, Name: "fun0", SystemName: "fun0", Filename: "file0"},
+		{ID: 2, Name: "fun1", SystemName: "fun1", Filename: "file1"},
+		{ID: 3, Name: "fun2", SystemName: "fun2", Filename: "file2"},
+		{ID: 4, Name: "fun3", SystemName: "fun3", Filename: "file3"},
+		{ID: 5, Name: "fun4", SystemName: "fun4", Filename: "file4"},
+		{ID: 6, Name: "fun5", SystemName: "fun5", Filename: "file5"},
+		{ID: 7, Name: "fun6", SystemName: "fun6", Filename: "file6"},
+		{ID: 8, Name: "fun7", SystemName: "fun7", Filename: "file7"},
+		{ID: 9, Name: "fun8", SystemName: "fun8", Filename: "file8"},
+		{ID: 10, Name: "fun9", SystemName: "fun9", Filename: "file9"},
+		{ID: 11, Name: "fun10", SystemName: "fun10", Filename: "file10"},
+	}
+	locations := []*Location{
+		{ID: 1, Mapping: mappings[0], Address: 0x1000, Line: []Line{{Function: functions[0], Line: 1}}},
+		{ID: 2, Mapping: mappings[0], Address: 0x2000, Line: []Line{{Function: functions[1], Line: 1}}},
+		{ID: 3, Mapping: mappings[0], Address: 0x3000, Line: []Line{{Function: functions[2], Line: 1}}},
+		{ID: 4, Mapping: mappings[0], Address: 0x4000, Line: []Line{{Function: functions[3], Line: 1}}},
+		{ID: 5, Mapping: mappings[0], Address: 0x5000, Line: []Line{{Function: functions[4], Line: 1}}},
+		{ID: 6, Mapping: mappings[0], Address: 0x6000, Line: []Line{{Function: functions[5], Line: 1}}},
+		{ID: 7, Mapping: mappings[0], Address: 0x7000, Line: []Line{{Function: functions[6], Line: 1}}},
+		{ID: 8, Mapping: mappings[0], Address: 0x8000, Line: []Line{{Function: functions[7], Line: 1}}},
+		{ID: 9, Mapping: mappings[0], Address: 0x9000, Line: []Line{{Function: functions[8], Line: 1}}},
+		{ID: 10, Mapping: mappings[0], Address: 0x10000, Line: []Line{{Function: functions[9], Line: 1}}},
+		{ID: 11, Mapping: mappings[1], Address: 0x11000, Line: []Line{{Function: functions[10], Line: 1}}},
+	}
+	profile := &Profile{
```

Comment:
In test_profile.go, we declare profiles (and functions/locations, etc) outside of the test cases so they can be copied and used in multiple tests. This may be something you want to consider doing.

Reply:
I put it in the function to avoid polluting the global namespace while it's only used by one function. we can always pull it out if we decide to reuse it.